// eth-saddle/dist/trace/descriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.augmentLogs = void 0;
const log_1 = require("./log");
const operator_1 = require("./operator");
const utils_1 = require("./utils");
function augmentLogs(logs, constants) {
    const inv = Object.entries(constants).reduce((inv, [k, v]) => {
        const key = utils_1.pad(utils_1.trim0x(v)).toLowerCase();
        return {
            ...inv,
            [key]: [
                {
                    operation: k,
                    inputs: []
                }
            ]
        };
    }, {});
    return logs.reduce(({ logs, info }, log, i, allLogs) => {
        const nextLog = allLogs[i + 1];
        const nextStack = nextLog ? nextLog.stack : undefined;
        let { extended, info: nInfo } = operator_1.describeOperation(log, nextStack || [], info);
        const nLog = new log_1.Log(log, extended, info.lastLog);
        return {
            logs: [...logs, nLog],
            info: {
                ...nInfo,
                lastLog: nLog
            }
        };
    }, { logs: [], info: { inv, sha: {} } });
}
exports.augmentLogs = augmentLogs;
