// eth-saddle/dist/trace/log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = void 0;
const chalk_1 = __importDefault(require("chalk"));
class Log {
    constructor(log, extended, lastLog) {
        this.depth = log.depth;
        this.error = log.error;
        this.gas = log.gas;
        this.gasCost = log.gasCost;
        this.memory = log.memory;
        this.op = log.op;
        this.pc = log.pc;
        this.stack = log.stack;
        this.storage = log.storage;
        this.desc = extended.desc;
        this.inputs = extended.inputs;
        this.outputs = extended.outputs;
        this.lastLog = lastLog;
    }
    setSource(source, sourceLine) {
        this.source = source;
        this.sourceLine = sourceLine;
    }
    setContract(contract, delegateCallDepth) {
        this.contract = contract;
        this.delegateCallDepth = delegateCallDepth;
    }
    show() {
        if (this.source) {
            console.log([
                `${chalk_1.default.gray("Log")}       pc=${this.pc} op=${this.op} source=${this.sourceLine}`,
                `Solidity  ${this.source.trim()}`,
                `          ↓`,
                `EVM Asm   ${chalk_1.default.blueBright(this.desc)}`,
                ``
            ].join("\n"));
        }
        else {
            console.log(chalk_1.default.blueBright(this.desc));
        }
    }
}
exports.Log = Log;
