// eth-saddle/dist/trace/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pad = exports.trim0x = exports.trimZero = exports.chunk = exports.lastN = exports.last = void 0;
function last(arr, n = 1) {
    return arr[arr.length - n];
}
exports.last = last;
function lastN(arr, n = 1) {
    return arr.slice(arr.length - n);
}
exports.lastN = lastN;
function chunk(arr, size) {
    let i = 0;
    let chunks = [];
    while (i < arr.length) {
        chunks.push(arr.slice(i, i + size));
        i += size;
    }
    return chunks;
}
exports.chunk = chunk;
function trimZero(num) {
    let i = 0;
    while (num[i] === '0') {
        i++;
    }
    return num.slice(i);
}
exports.trimZero = trimZero;
function trim0x(num) {
    if (num.slice(0, 2) === '0x') {
        return num.slice(2);
    }
    else {
        return num;
    }
}
exports.trim0x = trim0x;
function pad(val, align = 32) {
    let padLength = align - (val.length % align);
    return [...Array(padLength).keys()].map((_) => '0').join('') + val;
}
exports.pad = pad;
